ssl/ssl_asn1.c

```diff
@@ -356,10 +356,16 @@ SSL_SESSION *d2i_SSL_SESSION(SSL_SESSION **a, const unsigned char **pp,
     ret->ext.max_early_data = as->max_early_data;
 
     if (as->alpn_selected != NULL) {
-        if (!ssl_session_strndup((char **)&ret->ext.alpn_selected,
-                                 as->alpn_selected))
+        if (as->alpn_selected->length > 0) {
+            OPENSSL_free(ret->ext.alpn_selected);
+            ret->ext.alpn_selected = OPENSSL_malloc(as->alpn_selected->length);
+            if (ret->ext.alpn_selected == NULL)
+                goto err;
+        }
+        if (!ssl_session_memcpy(ret->ext.alpn_selected,
+                                &ret->ext.alpn_selected_len,
+                                as->alpn_selected, as->alpn_selected->length))
```

Comment:
ssl_session_memcpy essentially checks if as->alpn_selected is non-NULL, performs some sanity checks as->alpn_selected->length and finally memcpy. It should be noted that all these checks are performed above. as->alpn_selected is checked for being non-NULL, and as->alpn_selected->length is checked for positivity. In other words ssl_session_memcpy can be replaced with just memcpy and ret->ext.alpn_selected_len assigment. And so a bunch of redundant conditional branches gets eliminated. I suppose I have to apologize for not being able to switch off optimizing mode :-) As one can as well argue that I don't recognize that calling ssl_session_memcpy is more modular and hence more readable. But in such case why isn't there ssl_session_memdup? Kind of side-note-ish, so feel free to disregard.

Here is another question, more important. ssl_session_strndup actually worked differently from what's suggested here. How different? ssl_session_strndup was actually adding terminating nul character to the input. And question is if there is any code that is executed at some later point some code would **assume** that it's dealing with nul-terminated string (and doesn't take ->length into account)? Because if there is, then after this change there won't be enforced nul character, and such code would go out of buffer bounds. Is this valid concern?

Reply:
> As one can as well argue that I don't recognize that calling ssl_session_memcpy is more modular and hence more readable. But in such case why isn't there ssl_session_memdup? Kind of side-note-ish, so feel free to disregard.

I realised that now we're not using strndup it could be simplified even further along the lines as was already being done for ext.tick and ticket_appdata - so I copied that approach.

> Here is another question, more important. ssl_session_strndup actually worked differently from what's suggested here. How different? ssl_session_strndup was actually adding terminating nul character to the input. And question is if there is any code that is executed at some later point some code that would assume that it's dealing with nul-terminated string (and doesn't take ->length into account)? Because if there is, then after this change there won't be enforced nul character, and such code would go out of buffer bounds. Is this valid concern?

I reviewed all the usage of alpn_selected to double check. I didn't find any instance of it relying on being NUL-terminated, with the exception that I *did* find another instance of the same problem that I'm solving here - so I fixed that too.